adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala

```diff
@@ -735,41 +946,210 @@ trait GenomicRDD[T, U <: GenomicRDD[T, U]] {
     optPartitions: Option[Int] = None)(
       implicit tTag: ClassTag[T], xTag: ClassTag[X]): GenomicRDD[(Option[T], Iterable[X]), Z] = RightOuterShuffleJoinAndGroupByLeft.time {
 
-    val (partitionSize, endSequences) = joinPartitionSizeAndSequences(optPartitions,
-      genomicRdd)
+    val preparedForShuffleRegionJoin =
+      prepareForShuffleRegionJoin(genomicRdd, optPartitions)
+
+    val (leftRddToJoin, rightRddToJoin) =
+      (preparedForShuffleRegionJoin.leftRddToJoin,
+        preparedForShuffleRegionJoin.rightRddToJoin)
+
+    // what sequences do we wind up with at the end?
+    val combinedSequences = sequences ++ genomicRdd.sequences
 
-    // key the RDDs and join
     GenericGenomicRDD[(Option[T], Iterable[X])](
-      RightOuterShuffleRegionJoinAndGroupByLeft[T, X](endSequences,
-        partitionSize,
-        rdd.context).partitionAndJoin(flattenRddByRegions(),
-          genomicRdd.flattenRddByRegions()),
-      endSequences,
+      RightOuterShuffleRegionJoinAndGroupByLeft[T, X](leftRddToJoin, rightRddToJoin)
+        .partitionAndJoin(),
+      combinedSequences,
       kv => {
         (kv._2.flatMap(v => genomicRdd.getReferenceRegions(v)) ++
           kv._1.toSeq.flatMap(v => getReferenceRegions(v))).toSeq
+      }).asInstanceOf[GenomicRDD[(Option[T], Iterable[X]), Z]]
+  }
+
+  /**
+   * Copartitions two RDDs according to their ReferenceRegions.
+   *
+   * @note This is best used under the condition that (repeatedly)
+   * repartitioning is more expensive than calculating the proper location
+   * of the records of this.rdd. It requires a pass through the co-located
+   * RDD to get the correct partition(s) for each record. It will assign a
+   * record to multiple partitions if necessary.
+   * @param rddToCoPartitionWith The rdd to copartition to.
+   * @return The newly repartitioned rdd.
+   */
+  private[rdd] def copartitionByReferenceRegion[X, Y <: GenomicRDD[X, Y]](
```

Comment:
This could also be in ShuffleRegionJoin, but again, I can see it being in either place. At a higher level we should ask if we want the `set-theory` code to be the only place you can copartition two RDDs.